### DISCUSSION TITLE
What is the expected CRL caching / certificate selection behaviour in strongswan 5.9.5?

### DISCUSSION THREAD

**Discussion by stevelynchcp** [strongswan/strongswan#2039] [Q&A] [answered]
When I go through the following steps
- setup a CA using easyrsa. 
- generate a certificate for the initiator and responder and a blank CRL.
- copy the responder cert and CA cert to the responder and start strongswan.
- copy the initiator cert and CA cert to the initiator and start strongswan and initiate the connection
- The tunnel is up and I can ping across it.
- revoke both the initiator and responder certificates
- copy the new CRL to the initiator and responder
- run swanctl --load-all on the initator and responder. This shows the timestamp of the new CRL when loading the CRL
- run `swanctl --rekey --ike initiator --reauth` on the initiator.

I expect to see an auth failure because the certificates have been revoked and a CRL with the new timestamp in the log

I actually see a successful auth with the timestamp of the old CRL in the log.
running `swanctl --flush-certs` can work around this problem but it seems like very non obvious behaviour

I am also seeing strange behaviour if I copy a new certificate into place so I have two certificates with the same CN loaded. One revoked with the reason "superseded" and one that has not been revoked. The revoked certificate seems to be selected when doing re-authentication and the auth fails when it should succeed. In this case flushing the certificate cache doesn't seem to help.

I'm trying to work out the sequence of steps to go through to automatically rotate a certificate on an embedded machine with minimal downtime and no direct synchronisation between the devices so having both an old and new certificate in place for a short period of time would be necessary.

Is the behaviour I'm seeing the expected behaviour or is this a bug I should file?

I've included logs for the revocation condition and an example setup using docker-compose and shell scripts to show the behaviour I'm seeing.
[docker-compose-log.txt](https://github.com/strongswan/strongswan/files/13705527/docker-compose-log.txt)
[cert-renewal.tar.gz](https://github.com/strongswan/strongswan/files/13705685/cert-renewal.tar.gz)

Edit: Further investigation shows `swanctl --rekey --ike initiator --reauth` and `swanctl --terminate ... && swanctl --initiate ...` have different behaviour. terminate and initate seems to pick up a new certificate where reauth does not even after flushing the certs.

**Comment by strongX509** [comment]
* First **`swanctl`** doesn't have a **`--reauth`**  option.
* Second, as long as a CRL hasn't reached its **`nextUpdate`** deadline, the information contained in it is regarded as valid and no newer CRL is evaluated. If you you want to react quickly to a revoked certificate you have to either reduce the lifetime of a CRL or use OCSP instead.

**Reply by stevelynchcp** [reply]
Thanks for your reply.

With regard to point 1 
`swanctl --rekey`  does have a --reauth option

```
root@2efb67a8dd67:/app# swanctl --rekey --help
strongSwan 5.9.5 swanctl
usage:
  swanctl --rekey --child <name> | --ike <name> | --child-id <id> | --ike-id <id>
                 [--reauth] [--raw|--pretty]
...
           --reauth          (-a)  reauthenticate instead 
```


With regard to your second point
I'm loading a CRL from a file in the /etc/swanctl/x509crl directory. I'm surprised a newer update to that file is not evaluated if it's copied to that directory and loaded. It is detected as being newer when `swanctl --load-all` is run but not used when the certificate is checked.

If this is the designed behaviour then I find it confusing due to the notification that it has been loaded in the log. If this is working as designed then fair enough I can flush the cache to use the newly loaded CRL.

**Reply by stevelynchcp** [reply]
Sorry. I'm not sure this fell through the gaps over the holiday period or if it's considered working as designed.

**Comment by tobiasbrunner** [comment, marked as answer]
> I expect to see an auth failure because the certificates have been revoked and a CRL with the new timestamp in the log

Not if the in-memory cache is not flushed, because, as Andreas wrote, the existing CRL is still considered valid. The cache is not flushed automatically even if you load the CRL explicitly via `swanctl`. So a `--flush-certs` call is necessary after loading a new CRL (optionally pass  `--type x509_crl` to only flush CRLs).


> Further investigation shows `swanctl --rekey --ike initiator --reauth` and `swanctl --terminate ... && swanctl --initiate ...` have different behaviour. terminate and initate seems to pick up a new certificate where reauth does not even after flushing the certs.

Yes, a reauthenticaiton uses the (immutable and refcounted) connection objects of the existing connection, which will reference the old certificate, while the `--initiate` call will use the new connection objects that reference the new certificate. To avoid that, don't reference the certificate explicitly in the connection but just load it (which happens already if it's in the `x509` dir) and reference it via identity.

**Reply by stevelynchcp** [reply]
That worked. Thanks for clearing that up.
